libcloud/storage/drivers/runabove.py

```diff
@@ -0,0 +1,289 @@
+        """
+        action = '%s/storage/%s' % (API_ROOT, container_name)
+        location = self._check_location(ex_location)
+        data = {
+            'region': location,
+        }
+        try:
+            response = self.connection.request(action, data=data)
+        except ProviderError as err:
+            if err.http_code == httplib.NOT_FOUND:
+                raise ContainerDoesNotExistError(value=None,
+                                                 driver=self,
+                                                 container_name=container_name)
+            raise err
+        return self._to_container(response.object, location=location)
+
+    def get_object(self, container_name, object_name, ex_location=None):
+        # TODO: Remake with Swift request
+        location = self._check_location(ex_location)
+        action = '%s/storage/%s' % (API_ROOT, container_name)
+        data = {
+            'region': location,
+            'limit': 1000,
+        }
+        try:
+            response = self.connection.request(action, data=data)
+            json_response = response.object
+            container = self._to_container(json_response, location)
+            obj = [o for o in json_response['objects']
+                   if o['name'] == object_name][0]
+        except ProviderError as err:
+            if err.http_code == httplib.NOT_FOUND:
+                raise ContainerDoesNotExistError(value=None,
+                                                 driver=self,
+                                                 container_name=container_name)
+            raise err
+        except IndexError:
+            raise ObjectDoesNotExistError(object_name=object_name,
+                                          value='', driver=self)
+        return self._to_object(obj, container=container)
+
+    def _put_data(self, location, action, data):
+        """Common method for put data and except errors."""
+        conn = self._get_container_connection(location)
+        try:
+            conn.request(action=action, data=data, method="PUT")
+        except ProviderError as err:
+            if err.http_code == httplib.NOT_FOUND:
+                raise ContainerDoesNotExistError(value=None,
+                                                 driver=self,
+                                                 container_name=None)
+            raise err
+
+    def upload_object(self, file_path, container, object_name, extra=None,
+                      verify_hash=True, headers=None):
+        action = '/%s/%s' % (container.name, object_name)
+        data = open(file_path, 'rb').read()
+        self._put_data(container.extra['region'], action, data)
+        return container.get_object(object_name)
+
+    def upload_object_via_stream(self, iterator, container, object_name,
+                                 extra=None, headers=None):
+        action = '/%s/%s' % (container.name, object_name)
+        data = iterator.read()
```

Comment:
Hm, so this works around the whole "memory efficient" streaming approach implemented in the base driver...

This is less than ideal since, among other things, it results in reading and buffering the whole object in memory. Can you look at other drivers for an example (e.g. CloudFiles or S3, notably `upload_object`, `upload_object_via_stream`, `_upload_object_part` and `_put_object` methods) and see if you can use the same approach / pattern here?